src/app/trips/trip-forms/trip-form/trip-form.component.ts

```diff
@@ -0,0 +1,387 @@
+  // I'll try to fix this implementation in the near future
+  styleDivUnchangeable(fieldName: string, i: number) {
+    // make input DISABLED and remove bottom border
+    const fieldInput = document.getElementById(fieldName + i)!;
+    fieldInput.setAttribute('disabled', 'disabled');
+    fieldInput.style.borderBottomColor = 'transparent';
+
+    // change div background and border of field
+    let field = document.getElementById(fieldName + 'Field' + i)!;
+    field.style.backgroundColor = '#e4e4f4';
+    field.style.borderRadius = '10px';
+  }
+
+  // if user entered valid location but forgot to hit "add", add it for them on submit (to saved destinations objects [])
+  checkLastDestination() {
+    if (
+      this.destinations.length >
+      this.destinationsService.savedDestinations.length
+    ) {
+      this.destinationsService.saveDestination();
+    }
+  }
+
+  // Because adding members is optional, check and remove last value if it's blank
+  // this implementation will be changed and improved in the future
+  checkLastMember() {
+    let lastMemberIndex = this.members.length - 1;
+    let lastMemberControl = 'members.' + lastMemberIndex;
+    if (this.tripForm.get(lastMemberControl)?.value === '')
+      this.members.removeAt(lastMemberIndex);
+  }
+
+  // submit method
+  onSubmit() {
+    console.log('Submit button pushed!');
+    this.tripForm.markAllAsTouched();
+    if (this.tripForm.invalid) {
+      return false;
+    }
+
+    // double check that blank form array values aren't submitted,
+    // and valid but not 'added' values are still added
+    this.checkLastDestination();
+    this.checkLastMember();
+
+    // assign values from user to new trip object
+    this.setSubmittedTripValues();
+
+    if (this.isEditing) {
+      this.updateTrip();
+    } else {
+      this.addTrip();
+    }
+
+    console.log('Successfully saved!');
+    console.table(this.submittedTrip);
+    return true;
+  }
+
+  // add trip on submit of NEW trip form
+  addTrip() {
+    // new trip post request
+    this.tripService.createTrip(this.submittedTrip).subscribe(() => {
+      // deactivate route guard
+      this.tripComponent.isDirty = false;
+      //reroute to newly created trip
+      this.router.navigate([`/trips/${this.submittedTrip.id}`]);
+      this.destinationsService.clearAllDestinations();
+    });
+  }
+
+  // update trip on submit of EDIT trip form
+  updateTrip() {
+    // new trip put request
+    this.tripService.updateTrip(this.submittedTrip).subscribe(() => {
+      // deactivate route guard
+      this.tripComponent.isDirty = false;
+      //reroute to newly created trip
+      this.router.navigate([`/trips/${this.submittedTrip.id}`]);
+      this.destinationsService.clearAllDestinations();
+    });
+  }
+
+  // cancel method
+  cancel() {
+    this.router.navigate(['/trips']);
+  }
+
+  deleteTrip() {
+    // request confirmation
+    if (window.confirm(`Are you sure you would like to permanently delete this trip?`)) {
```

Comment:
we want to avoid `window` as well. In this case I don't think you need the keyword window to use confirm. You may need to check on that though.